Use new API for reading nodes in interpreted runtime

pontusmelke opened the pull request:
This PR switches to use the new API in `QueryContex`t for `all`, `allPrimitive` and label and property reads. The PR also contains a refactoring of `QueryContext` so that operations return `NodeValue` and `EdgeValue` instead of `Node` and `Relationship`.

~Builds on top of https://github.com/neo4j/neo4j/pull/10495~

pontusmelke committed:
Added new utilities to tx state

- means to keep track of changes for a property reference
- minor utility method to find changed properties

pontusmelke committed:
Move LabelNotFoundException to API module

pontusmelke committed:
Add labelGetName to Token

pontusmelke committed:
Switch to new API for getLabelsForNode

pontusmelke committed:
Removed unused method from QueryContext

pontusmelke committed:
Resurrect ResourceManager from the gits

pontusmelke committed:
Add ability to track resources in QueryContext

pontusmelke committed:
Refactor: QueryContext returns NodeValue and EdgeValue

pontusmelke committed:
Use new api for allNode scans

pontusmelke committed:
Use new API for property reads

pontusmelke committed:
Use new API for more label operations

pontusmelke committed:
Some clean up after reviewing

pontusmelke committed:
Added tests for cursor closing

fickludd commented on community/cypher/interpreted-runtime/src/main/scala/org/neo4j/cypher/internal/runtime/interpreted/TransactionBoundQueryContext.scala:
```diff
@@ -161,11 +161,17 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+      reads().singleNode(node, nodeCursor)
+      if (!nodeCursor.next()) false
+      else nodeCursor.labels().contains(label)
+    }
```
Forgot to close the cursor. If we go this route, we will not some utility for try-with-resources in scala.

pontusmelke replied:
we already have that https://github.com/neo4j/neo4j/blob/3.4/community/cypher/frontend-3.4/src/main/scala/org/neo4j/cypher/internal/frontend/v3_4/helpers/package.scala#L27. However requires lambda, thought we'll do it without allocation here but you are right that I forgot to close it here.

fickludd replied:
1) Cool, I didn't know we had that... should be moved out of the frontend module though.
2) Ok, I like less allocation, so let's to that.

pontusmelke committed:
Close cursor

pontusmelke committed:
hasChanges -> hasPropertyChanges and hasRelatsionshipChanges

pontusmelke committed:
Fixed embarrassing bug

pontusmelke committed:
Rebasing conflicts

pontusmelke committed:
NodeCursor must be stable

For queries like `MATCH () CREATE ()` we need the cursor to be stable and
only check the initial tx state.

pontusmelke committed:
Fix failing tests

pontusmelke committed:
Fix rebase issues

pontusmelke committed:
Add snapshot of added nodes in tx state

For queries like `MATCH () CREATE ()` we need to keep a snapshot of the
added nodes in the transaction state and not just endlessly update with
nodes added after the initial MATCH.

pontusmelke committed:
Fix failed tests due to mocking problems.

fickludd commented:
I'm a little surprised by the added-nodes snapshot in `NodeCursor`. Did we do this before? I though it was only needed to freeze the `hasTxStateWithChanges()` check on init... I guess that breaks on `CREATE ... MATCH () CREATE ()`. Mystified!

pontusmelke commented:
@fickludd No we just thought/hoped we could do without it but we ended up with endless loops for stuff like

```
try(Transaction tx = db.beginTx())
{
   db.execute("CREATE ()");
   db.execute("MATCH() CREATE()");
}
```
In this case we will have `hasTxStateWithChanges == true` and then we will be endlessly matching on changes in the tx state and create new nodes until we run out of memory.
